cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go

```diff
@@ -370,9 +368,51 @@ func (p *BicepProvider) Destroy(
 		})
 }
 
-func (p *BicepProvider) getResourceGroupsFromDeployment(ctx context.Context) ([]string, error) {
+// latestDeployment finds the most recent deployment for the given environment in the provided scope.
+func (p *BicepProvider) latestDeployment(ctx context.Context, scope infra.Scope) (*armresources.DeploymentExtended, error) {
+	deployments, err := scope.ListDeployments(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	slices.SortFunc(deployments, func(x, y *armresources.DeploymentExtended) bool {
+		return x.Properties.Timestamp.After(*y.Properties.Timestamp)
+	})
+
+	// Earlier versions of `azd` did not use unique deployment names per deployment and also did not tag the deployment
+	// with an `azd` specific tag. Instead, the name of the deployment simply matched the environment name.
+	//
+	// As we walk the list of deployments, we note if we find a deployment matching this older strategy and will return
+	// it if we can't find a deployment that matches the newer one.
+	var matchingBareDeployment *armresources.DeploymentExtended
+
+	for _, deployment := range deployments {
+		if v, has := deployment.Tags["azd-env-name"]; has && *v == p.env.GetEnvName() {
+			return deployment, nil
+		}
+
+		if *deployment.Name == p.env.GetEnvName() {
+			matchingBareDeployment = deployment
+		}
+	}
+
+	if matchingBareDeployment != nil {
+		return matchingBareDeployment, nil
+	}
+
+	return nil, fmt.Errorf("no deployment found for environment %s", p.env.GetEnvName())
+}
+
+func (p *BicepProvider) getResourceGroups(ctx context.Context) ([]string, error) {
+	scope := infra.NewSubscriptionScope(p.azCli, p.env.GetSubscriptionId())
+	latestDeployment, err := p.latestDeployment(ctx, scope)
+	if err != nil {
+		return []string{}, err
+	}
+
 	resourceManager := infra.NewAzureResourceManager(p.azCli)
-	resourceGroups, err := resourceManager.GetResourceGroupsForDeployment(ctx, p.env.GetSubscriptionId(), p.env.GetEnvName())
+	resourceGroups, err := resourceManager.GetResourceGroupsForDeployment(
```

Comment:
There is an optimization opportunity here, I think given that we already fetched deployment contents, but `GetResourceGroupsForDeployment` will do that again.  Happy to pull that into here, or I can file an issue to track it with a follow up.

Reply:
I don't mind giving this the one-pass if you want to remove `GetResourceGroupsForDeployment` and simplify the logic here.

I worry less about the extra http request, more that `GetResourceGroupsForDeployment` has rather loose behavior after looking at it again. It's not clear what or how resource groups are determined in the method (I completely missed this when making an earlier change), or how subscription scope deployment plays into this.

Reply:
Pulled the logic from GetResourceGroupsForDeployment into here (this was the only caller) and updated it to work on the discovered deployment we fetched earlier.  I like this much better, thanks for the nudge.